cabal-validate/src/Main.hs

```diff
@@ -0,0 +1,430 @@
+      <> show (verbose opts)
+      <> "\nextra compilers:   "
+      <> unwords (extraCompilers opts)
+      <> "\nextra RTS options: "
+      <> unwords (rtsArgs opts)
+
+-- | Print the versions of tools being used.
+printToolVersions :: Opts -> IO ()
+printToolVersions opts = do
+  timed opts (compilerExecutable (compiler opts)) ["--version"]
+  timed opts (cabal opts) ["--version"]
+
+  forM_ (extraCompilers opts) $ \compiler' -> do
+    timed opts compiler' ["--version"]
+
+-- | Run the build step.
+build :: Opts -> IO ()
+build opts = do
+  printHeader "build (dry run)"
+  timed
+    opts
+    (cabal opts)
+    ( cabalNewBuildArgs opts
+        ++ targets opts
+        ++ ["--dry-run"]
+    )
+
+  printHeader "build (full build plan; cached and to-be-built dependencies)"
+  timed
+    opts
+    "jq"
+    [ "-r"
+    , -- TODO: Maybe use `cabal-plan`? It's a heavy dependency though...
+      ".\"install-plan\" | map(.\"pkg-name\" + \"-\" + .\"pkg-version\" + \" \" + .\"component-name\") | join(\"\n\")"
+    , baseBuildDir opts </> "cache" </> "plan.json"
+    ]
+
+  printHeader "build (actual build)"
+  timed
+    opts
+    (cabal opts)
+    (cabalNewBuildArgs opts ++ targets opts)
+
+-- | Run doctests.
+--
+-- This doesn't work on my machine, maybe @cabal.nix@ needs some love to
+-- support @cabal-env@?
+doctest :: Opts -> IO ()
+doctest opts = do
+  timed
+    opts
+    "cabal-env"
+    [ "--name"
+    , "doctest-cabal"
+    , "--transitive"
+    , "QuickCheck"
+    ]
+
+  timed
+    opts
+    "cabal-env"
+    [ "--name"
+    , "doctest-cabal"
+    , "array"
+    , "bytestring"
+    , "containers"
+    , "deepseq"
+    , "directory"
+    , "filepath"
+    , "pretty"
+    , "process"
+    , "time"
+    , "binary"
+    , "unix"
+    , "text"
+    , "parsec"
+    , "mtl"
+    ]
+
+  timed
+    opts
+    "doctest"
+    [ "-package-env=doctest-Cabal"
+    , "--fast"
+    , "Cabal/Distribution"
+    , "Cabal/Language"
+    ]
+
+-- | Run tests for the @Cabal@ library, and also `runHackageTests` if those are
+-- enabled.
+libTests :: Opts -> IO ()
+libTests opts = do
+  let runCabalTests' suite extraArgs =
+        timedCabalBin
+          opts
+          "Cabal-tests"
+          ("test:" <> suite)
+          ( tastyArgs opts
+              ++ jobsArgs opts
+              ++ extraArgs
+          )
+
+      runCabalTests suite = runCabalTests' suite []
+
+  runCabalTests' "unit-tests" ["--with-ghc=" <> compilerExecutable (compiler opts)]
+  runCabalTests "check-tests"
+  runCabalTests "parser-tests"
+  runCabalTests "rpmvercmp"
+  runCabalTests "no-thunks-test"
+
+  runHackageTests opts
+
+-- | Run Hackage tests, if enabled.
+runHackageTests :: Opts -> IO ()
+runHackageTests opts
+  | NoHackageTests <- hackageTests opts = pure ()
+  | otherwise = do
+      command <- cabalListBin opts "Cabal-tests:test:hackage-tests"
+
+      let
+        -- See #10284 for why this value is pinned.
+        hackageTestsIndexState = "--index-state=2024-08-25"
```

Comment:
I really hope we can find a way to externalise this value at some point. Just a text file with the timestamp and a descriptive name, which cabal-validate could read in, perhaps?

Reply:
If we didn't have to deal with 3.10.3's broken `import`, that would be easy to do. As is, it's nightmarish to get right in a way that works with all cabal versions in use for Cabal development.